dynamallow/local.py

```diff
@@ -0,0 +1,80 @@
+import os
+import random
+import socket
+import tarfile
+import tempfile
+
+import subprocess
+
+try:
+    from urllib import urlretrieve
+except ImportError:
+    from urllib.request import urlretrieve
+
+
+class DynamoLocal(object):
+    """
+    Spins up a local dynamo instance. This should ONLY be used for testing!! You must call
+    ``dynamo_local.shutdown()`` when you are done with the instance.
+    """
+    def __init__(self, dynamo_dir, port=None, log=None):
+        self.port = port or self._get_random_port()
+        if not os.path.isdir(dynamo_dir):
+            if log:
+                log.info("Creating dynamo_local_dir: {0}".format(dynamo_dir))
+            assert not os.path.exists(dynamo_dir)
+            os.makedirs(dynamo_dir, 0o755)
+
+        if not os.path.exists(os.path.join(dynamo_dir, 'DynamoDBLocal.jar')):
+            temp_fd, temp_file = tempfile.mkstemp()
+            os.close(temp_fd)
+            if log:
+                log.info("Downloading dynamo local to: {0}".format(temp_file))
+            urlretrieve(
+                'http://dynamodb-local.s3-website-us-west-2.amazonaws.com/dynamodb_local_latest.tar.gz',
+                temp_file
+            )
+
+            log.info("Extracting dynamo local...")
+            archive = tarfile.open(temp_file, 'r:gz')
+            archive.extractall(dynamo_dir)
+            archive.close()
+
+            os.unlink(temp_file)
+
+        if log:
+            log.info("Running dynamo from {dir} on port {port}".format(
+                dir=dynamo_dir,
+                port=self.port
+            ))
+        self.dynamo_proc = subprocess.Popen(
+            (
+                'java',
+                '-Djava.library.path=./DynamoDBLocal_lib',
+                '-jar', 'DynamoDBLocal.jar',
+                '-sharedDb',
+                '-inMemory',
+                '-port', str(self.port)
+            ),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            cwd=dynamo_dir
+        )
+
+    def shutdown(self):
+        self.dynamo_proc.terminate()
+        self.dynamo_proc.wait()
+
+    @staticmethod
+    def _get_random_port():
```

Comment:
I see no reason to mark this as a "private" method.

In fact, I would simply move it to a regular method outside of the class since it really don't have anything to specifically do with running dynamo.